### DISCUSSION TITLE
The `IAccount` function `validateTransaction` is unclear to me. When/Why/How would we pass in the transaction hash of our transaction?

### DISCUSSION THREAD

**Discussion by PatrickAlphaC** [zkSync-Community-Hub/zksync-developers#487] [General]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

According to the [zkSync account abstraction IAccount interface](https://github.com/matter-labs/era-contracts/blob/main/system-contracts/contracts/interfaces/IAccount.sol), there is a function named `validateTransaction` who's definition is as such:

```javascript
    /// @notice Called by the bootloader to validate that an account agrees to process the transaction
    /// (and potentially pay for it).
    /// @param _txHash The hash of the transaction to be used in the explorer
    /// @param _suggestedSignedHash The hash of the transaction is signed by EOAs
    /// @param _transaction The transaction itself
    /// @return magic The magic value that should be equal to the signature of this function
    /// if the user agrees to proceed with the transaction.
    /// @dev The developer should strive to preserve as many steps as possible both for valid
    /// and invalid transactions as this very method is also used during the gas fee estimation
    /// (without some of the necessary data, e.g. signature).
    function validateTransaction(
        bytes32 _txHash,
        bytes32 _suggestedSignedHash,
        Transaction calldata _transaction
    ) external payable returns (bytes4 magic);
``` 

A few questions about this:
1. When this is referring to `transaction hash` by way of `txHash` or `hash of the transaction` are they referring to the literal transaction hash as the unique identifier of a transaction? Or a hash of the `Transaction` **object**, as seen below:

```javascript
struct Transaction {
    uint256 txType;
    uint256 from;
    uint256 to;
    uint256 gasLimit;
// etc..
}
```

2. If it's the unique identifier of the transaction, how would I know the `_txHash` beforehand? Why would that be important to validating the operation? (Note: _None_ of the tutorials from the docs make use of this parameter)
3. Why is there a _second_ parameter called `_suggestedSignedHash`? How would we know this? Why would we know this?

In the custom-aa-tutorial, we see our function called like so:
```javascript
    function executeTransactionFromOutside(Transaction calldata _transaction)
        external
        payable
    {
        _validateTransaction(bytes32(0), _transaction); // <----- this line
        _executeTransaction(_transaction);
    }
```

Where we pass `bytes32(0)` as the `_suggestedSignedHash`. Why do we even have those? What do these do? 
4. Why did we call the `Transaction` struct `Transaction`? It now makes it confusing if we are referring to a `Transaction` like the transaction or a `Transaction` like the struct? 

### What is the specific issue or error you're encountering?

See above

### Can you share the error messages or logs you're receiving, if any?

See above

### Have you made any recent changes to the contract before encountering this issue?

See above

### Are there any external libraries or contracts that your contract interacts with?

See above

### Can you provide the relevant portions of your contract code where the issue is occurring?

See above

### Have you tried to isolate the problem, and if so, what were the results?

See above

### What steps have you already taken to try to resolve the issue?

See above

**Comment by bxpana** [comment]
Thanks @PatrickAlphaC we'll take some time to check this out and get back to you.

**Comment by AlbionaHoti** [comment]
Hi @PatrickAlphaC,

Here's an overview of what I could collect:

1. gets the transaction itself _transaction and asks whether the account should pay for it later. The default account would pay for the transaction based on some basic parameters, such as
    - only if the signature is valid
    - if the nonce is correct
    - this function makes the authorization; it can override the authorization by changing the signature scheme or playing with nonce behavior
    - this function checks the transaction parameters basically
1.1 The validateTransaction function is called by the bootloader → shrunk into the protocol; only the protocol can call the two functions: `validateTransaction` and `executeTransaction.`

- The `executeTransaction` function processes the transaction that was already validated and proceeds with the execution

- `_suggestedSignature` →
    - If the Bootloader/system specifies it, you can use it; if not, we can use a library called: `.encodeHash` → to get the transaction hash. → getting all the fields together, adding some salt and hash, and having a unique identifier.
    
For:

> 1. Where we pass bytes32(0) as the _suggestedSignedHash. Why do we even have those? What do these do? 

I'll come back with a clearer and fuller response!
